packages/block-library/src/query/index.js

```diff
@@ -64,4 +71,58 @@ export const settings = {
 	deprecated,
 };
 
+let previousPostTypes = [];
+let haltSubscribing = false;
+
+const keywords = {
+	post: [ 'blog' ],
+};
+
+subscribe( () => {
+	if ( haltSubscribing ) {
+		return;
+	}
+	const { getPostTypes } = select( coreStore );
+	const excludedPostTypes = [ 'attachment' ];
+	const filteredPostTypes = ( getPostTypes( { per_page: -1 } ) ?? [] ).filter(
+		( { viewable, slug } ) =>
+			viewable && ! excludedPostTypes.includes( slug )
+	);
+
+	for ( const postType of previousPostTypes ) {
+		if ( ! filteredPostTypes.includes( postType ) ) {
+			haltSubscribing = true;
+			unregisterBlockVariation( name, postType.slug );
+			haltSubscribing = false;
+		}
+	}
+
+	for ( const postType of filteredPostTypes ) {
+		if ( ! previousPostTypes.includes( postType ) ) {
+			haltSubscribing = true;
+			registerBlockVariation( name, {
+				name: postType.slug,
+				title: sprintf(
+					/* translators: %s: post type */
+					__( '%s Query Loop' ),
+					postType.labels.singular_name
+				),
+				attributes: {
+					query: {
+						postType: postType.slug,
```

Comment:
I think this PR is going to create more issues that it solves, mainly because if you select a post type, you can't set the query to "inherit" true.

Reply:
There are more nuances than `inherit`. An example is the `isActive`, existing Posts List variation, Pages List block etc..

Reply:
Ok, then maybe we need to change the name depending on the main query that the current template is on.

Reply:
Also, there's already a "Posts List" variation, we should probably remove that then

Reply:
Removed it here: #63404

Reply:
If we'd remove the Posts List variation, it should be done here and definitely not in it's own PR without having the alternative landed..

Please see the issue I linked above for some more nuances.
Additionally as @Mamaduka also mentions, this should be done server side and avoid all this complexity in the client and the subscription.

Reply:
Also see #63353 description where I share some thoughts on this.

Reply:
I was going to close this PR because of the issues? Not sure if it makes sense to add the variations at all, if it has to inherit the query, we should probably dynamically name it based on what the main query is.